dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -0,0 +1,1068 @@
+/*
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#include "DynamicData.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace XTypes {
+
+DDS::ReturnCode_t DynamicData::get_descriptor(MemberDescriptor& value, MemberId id) const
+{
+}
+
+DDS::ReturnCode_t DynamicData::set_descriptor(MemberId id, const MemberDescriptor& value)
+{
+}
+
+bool DynamicData::equals(const DynamicData& other) const
+{
+}
+
+MemberId DynamicData::get_member_id_by_name(String name) const
+{
+}
+
+MemberId DynamicData::get_member_id_at_index(ACE_CDR::ULong) const
+{
+}
+
+ACE_CDR::ULong DynamicData::get_item_count() const
+{
+}
+
+DDS::ReturnCode_t DynamicData::clear_all_values()
+{
+}
+
+DDS::ReturnCode_t DynamicData::clear_nonkey_values()
+{
+}
+
+DDS::ReturnCode_t DynamicData::clear_value(MemberId id)
+{
+}
+
+DynamicData DynamicData::loan_value(MemberId id)
+{
+}
+
+DDS::ReturnCode_t DynamicData::return_loaned_value(const DynamicData& value)
+{
+}
+
+DynamicData DynamicData::clone() const
+{
+}
+
+DDS::ReturnCode_t DynamicData::get_int32_value(ACE_CDR::Long& value, MemberId id) const
+{
+  const OPENDDS_MAP<MemberId, size_t>::const_iterator it = offset_lookup_table_.find(id);
+  if (it != offset_lookup_table_.cend()) {
+    size_t offset = it->second;
+    std::ostringstream err_msg;
+    err_msg << "Failed to skip to the member with ID " << id;
+    if (!skip("get_int32_value", err_msg.str().c_str(), offset)) {
+      return DDS::RETCODE_ERROR;
+    }
+
+    if (!(strm_ >> value)) {
+      if (DCPS_debug_level >= 10) {
```

Comment:
The guideline in debug.h says it should be 10 if the logs may happen more than once per sample. It seems to be the case for this class because it operates on a single sample and can generate multiple logs.

Reply:
The list in debug.h isn't a complete guide because it doesn't mention errors at all. I think it assumes errors are unconditionally logged as that is the case in a lot of the older code and it puts warnings at 1. So I don't think that list really applies here because errors should be higher priority than normal debug logging. For example I have [errors at 1 to 3 in the security logging scheme](https://github.com/objectcomputing/OpenDDS/blob/1fd77cf02fd6f3ed9f1fb3c873455619816d4e75/dds/DCPS/debug.cpp#L107-L109). Thinking about it again, it probably doesn't have to be at level 1 if there is another log point down the line at 1 to report ultimately that an error happened. I don't have a concrete number to tell you, but theoretically if we wanted to put a normal debug statement in this function, then it'd have to have the same priority as an error, which doesn't seem right to me.

Also I'm not sure how this happens more than once per sample. If there an error in deserialization we won't come back here, right? So if this wasn't an actually error, I would say should be at debug level 8.

Reply:
> The list in debug.h isn't a complete guide because it doesn't mention errors at all. I think it assumes errors are unconditionally logged as that is the case in a lot of the older code and it puts warnings at 1. So I don't think that list really applies here because errors should be higher priority than normal debug logging. For example I have [errors at 1 to 3 in the security logging scheme](https://github.com/objectcomputing/OpenDDS/blob/1fd77cf02fd6f3ed9f1fb3c873455619816d4e75/dds/DCPS/debug.cpp#L107-L109). Thinking about it again, it probably doesn't have to be at level 1 if there is another log point down the line at 1 to report ultimately that an error happened. I don't have a concrete number to tell you, but theoretically if we wanted to put a normal debug statement in this function, then it'd have to have the same priority as an error, which doesn't seem right to me.
> 
Yeah, this makes sense. I can change this to log unconditionally since I think this is an error and should be logged regardless of the debug level.

> Also I'm not sure how this happens more than once per sample. If there an error in deserialization we won't come back here, right? So if this wasn't an actually error, I would say should be at debug level 8.
>
So a user can call the same method more than once even though there is an error in the first call. It doesn't make much sense if they do so, but we don't disallow that.

Reply:
> Yeah, this makes sense. I can change this to log unconditionally since I think this is an error and should be logged regardless of the debug level.

I meant that because we don't always unconditionally log errors anymore, the list in debug.h isn't completely representative of what our current practice is. If this error status is going to returned to the user, then it definitely shouldn't be logged unconditionally.